src/hotspot/share/opto/parse.hpp

```diff
@@ -603,4 +603,44 @@ class Parse : public GraphKit {
 #endif
 };
 
+// Specialized uncommon_trap of unstable_if, we have 2 optimizations for them:
+//   1. suppress trivial Unstable_If traps
+//   2. use next_bci of _path to update live locals.
+class UnstableIfTrap {
+  CallStaticJavaNode* const _unc;
+  // Parse::_blocks outlive Parse object itself.
+  // They are reclaimed by ResourceMark in CompileBroker::invoke_compiler_on_method().
+  Parse::Block* const _path; // the pruned path
```

Comment:
Do we really need to keep track of the entire Block? Looks like we could just save next_bci as int.

Reply:
good point. 
I used path() in the second patch.  now I store those traps in their basic blocks. I don't need to remember them.